bee-ledger/src/workers/pruning/condition.rs

```diff
@@ -1,49 +1,131 @@
 // Copyright 2020-2021 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
+// use std::time::Duration;
+
+use std::{
+    sync::atomic::{AtomicU64, Ordering},
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
 use bee_message::milestone::MilestoneIndex;
 use bee_tangle::{storage::StorageBackend, Tangle};
 
 use crate::{types::LedgerIndex, workers::pruning::config::PruningConfig};
 
 const PRUNING_BATCH_SIZE_MAX: u32 = 200;
 
+static PREVIOUS_PRUNING_BY_SIZE: AtomicU64 = AtomicU64::new(0);
+
 /// Reasons for skipping pruning.
-#[derive(Debug)]
-pub enum PruningSkipReason {
-    /// Pruning is disabled in the config.
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum PruningSkipReason {
+    #[error("disabled")]
     Disabled,
-    /// Not enough data yet to be pruned.
-    BelowThreshold { reached_in: u32 },
+    #[error("ledger index < target index threshold")]
+    BelowTargetIndexThreshold,
+    #[error("size metric not supported by the storage layer")]
+    SizeMetricUnsupported,
+    #[error("size metric currently unavailable")]
+    SizeMetricUnavailable,
+    #[error("current size < target size threshold")]
+    BelowTargetSizeThreshold,
+    #[error("cooldown")]
+    Cooldown,
+}
+
+pub(crate) enum PruningTask {
+    ByIndexRange {
+        start_index: MilestoneIndex,
+        target_index: MilestoneIndex,
+    },
+    ByDbSize {
+        num_bytes_to_prune: usize,
+    },
 }
 
-pub(crate) fn should_prune<B: StorageBackend>(
-    tangle: &Tangle<B>,
+pub(crate) fn should_prune<S: StorageBackend>(
+    tangle: &Tangle<S>,
+    storage: &S,
     ledger_index: LedgerIndex,
-    pruning_delay: u32,
-    config: &PruningConfig,
-) -> Result<(MilestoneIndex, MilestoneIndex), PruningSkipReason> {
-    if config.disabled() {
+    milestones_to_keep: u32,
+    pruning_config: &PruningConfig,
+) -> Result<PruningTask, PruningSkipReason> {
+    if !pruning_config.milestones().enabled() && !pruning_config.db_size().enabled() {
         return Err(PruningSkipReason::Disabled);
     }
 
-    let pruning_index = *tangle.get_pruning_index() + 1;
-    let pruning_threshold = pruning_index + pruning_delay;
+    let pruning_by_size = if pruning_config.db_size().enabled() {
+        let prev = Duration::from_secs(PREVIOUS_PRUNING_BY_SIZE.load(Ordering::Relaxed));
+        // Panic: should not cause problems on properly set up hosts.
+        let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
+
+        if now < prev + pruning_config.db_size().cooldown_time() {
+            Err(PruningSkipReason::Cooldown)
+        } else {
+            let actual_size = {
+                if let Ok(size) = storage.size() {
+                    if let Some(size) = size {
```

Comment:
maybe we can do `size.ok_or_else(|| ...)` instead.

Reply:
good catch. `size.ok_or` :+1: